### PR TITLE
[RyuJIT] Fix get_ManagedThreadId(get_CurrentThread) optimization (convert to NI)

### LINKED ISSUE #41212: [RyuJIT] Thread.CurrentThread.ManagedThreadId optimization question
I noticed that there is an optimization for `Thread.CurrentThread.ManagedThreadId` where it's supposed to replace both getters(?) with a helper (`CORINFO_HELP_GETCURRENTMANAGEDTHREADID`) here: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/morph.cpp#L9044-L9062

but I don't see it happening in the following code:
```csharp
static int CurrId() => Thread.CurrentThread.ManagedThreadId;
```
Codegen for `CurrId` (tier1, promoted from tier0 via a loop):
```asm
; Assembly listing for method Program:CurrThreadId():int
; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; Tier-1 compilation
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 OutArgs      [V00    ] (  1,  1   )  lclBlk (32) [rsp+0x00]   "OutgoingArgSpace"
;  V01 tmp1         [V01,T00] (  2,  4   )     ref  ->  rcx         class-hnd "dup spill"
;  V02 tmp2         [V02,T01] (  4,  3.25)     ref  ->  rcx
;
; Lcl frame size = 40

G_M22857_IG01:
       4883EC28             sub      rsp, 40
                                                ;; bbWeight=1    PerfScore 0.25
G_M22857_IG02:
       48B928000C8FFC7F0000 mov      rcx, 0x7FFC8F0C0028
       BA78020000           mov      edx, 632
       E838F7765F           call     CORINFO_HELP_GETSHARED_GCTHREADSTATIC_BASE
       488B4818             mov      rcx, gword ptr [rax+24]
       4885C9               test     rcx, rcx
       7508                 jne      SHORT G_M22857_IG04
                                                ;; bbWeight=1    PerfScore 4.75
G_M22857_IG03:
       E8EAA7FDFF           call     System.Threading.Thread:InitializeCurrentThread():System.Threading.Thread
       488BC8               mov      rcx, rax
                                                ;; bbWeight=0.25 PerfScore 0.31
G_M22857_IG04:
       3909                 cmp      dword ptr [rcx], ecx
       E8705B675F           call     System.Threading.Thread:get_ManagedThreadId():int:this
       90                   nop
                                                ;; bbWeight=1    PerfScore 3.25
G_M22857_IG05:
       4883C428             add      rsp, 40
       C3                   ret
                                                ;; bbWeight=1    PerfScore 1.25

; Total bytes of code 54, prolog size 4, PerfScore 15.21, (MethodHash=c3d9a6b6) for method Program:CurrThreadId():int
; ============================================================
```

@dotnet/jit-contrib @benaadams

category:cq
theme:jit-intrinsics
skill-level:intermediate
cost:small

### PR DESCRIPTION
**Pull request by EgorBo**
Fixes: https://github.com/dotnet/runtime/issues/41212 (see comment https://github.com/dotnet/runtime/issues/41212#issuecomment-678651520)
```csharp
static int CurrId() => Thread.CurrentThread.ManagedThreadId; // or Environment.CurrentManagedThreadId
```
Old codegen:
```asm
G_M22857_IG01:
       4883EC28             sub      rsp, 40
G_M22857_IG02:
       48B928000C8FFC7F0000 mov      rcx, 0x7FFC8F0C0028
       BA78020000           mov      edx, 632
       E838F7765F           call     CORINFO_HELP_GETSHARED_GCTHREADSTATIC_BASE
       488B4818             mov      rcx, gword ptr [rax+24]
       4885C9               test     rcx, rcx
       7508                 jne      SHORT G_M22857_IG04
G_M22857_IG03:
       E8EAA7FDFF           call     System.Threading.Thread:InitializeCurrentThread():System.Threading.Thread
       488BC8               mov      rcx, rax
G_M22857_IG04:
       3909                 cmp      dword ptr [rcx], ecx
       E8705B675F           call     System.Threading.Thread:get_ManagedThreadId():int:this
       90                   nop
G_M22857_IG05:
       4883C428             add      rsp, 40
       C3                   ret
; Total bytes of code 54, prolog size 4, PerfScore 15.21, (MethodHash=c3d9a6b6) for method Program:CurrThreadId():int
; ============================================================
```
New codegen:
```asm
G_M22857_IG01:
       4883EC28             sub      rsp, 40
G_M22857_IG02:
       E8C7D2765F           call     CORINFO_HELP_GETCURRENTMANAGEDTHREADID
       90                   nop
G_M22857_IG03:
       4883C428             add      rsp, 40
       C3                   ret

; Total bytes of code 15, prolog size 4, PerfScore 4.25, (MethodHash=c3d9a6b6) for method Program:CurrThreadId():int
; ============================================================
```

Not sure I removed `Thread:InitializeCurrentThread()` correctly, see comments.

### CODE REVIEW

**Comment by jkotas** [comment]
> Benchmark:
> ```csharp
[Benchmark]
public int CurrentThreadId() => Environment.CurrentManagedThreadId;
```
```
       |          Method |      Mean |     Error |    StdDev |
       |---------------- |----------:|----------:|----------:|
PR     | CurrentThreadId |  0.859 ns | 0.0075 ns | 0.0071 ns |
master | CurrentThreadId |  3.270 ns | 0.0163 ns | 0.0136 ns |
```

The performance of `Environment.CurrentManagedThreadId` should be fixed by converting it to FCall that just does the right thing. It does not make sense to depend on JIT optimizations to do this. It is how it was done in .NET Framework, but it was deoptimized in early .NET Core days when Environment did not live in CoreLib.

**Comment by EgorBo** [comment]
> > Benchmark:
> > [Benchmark]
> > public int CurrentThreadId() => Environment.CurrentManagedThreadId;
> 
> The performance of `Environment.CurrentManagedThreadId` should be fixed by converting it to FCall that just does the right thing. It does not make sense to depend on JIT optimizations to do this. It is how it was done in .NET Framework, but it was deoptimized in early .NET Core days when Environment did not live in CoreLib.

What about `Thread.CurrentThread.ManagedThreadId`?

e.g. aspnetcore repo:
`Thread.CurrentThread.ManagedThreadId` usages: 6
`Environment.CurrentManagedThreadId` usages: 0

I honestly never used `Environment.CurrentManagedThreadId` API. But your suggestion makes things a lot easier I agree.

**Comment by jkotas** [comment]
> What about Thread.CurrentThread.ManagedThreadId?

Yes, this optimization is still useful to fix `Thread.CurrentThread.ManagedThreadId`. It may be also worth having analyzer that suggests converting `Thread.CurrentThread.ManagedThreadId` to `Environment.CurrentManagedThreadId`.

**Comment by jkotas** [comment]
Fixed `Environment.CurrentManagedThreadId` in #41360

**Comment by BruceForstall** [comment]
@EgorBo Are you still working on this?

**Comment by EgorBo** [comment]
@BruceForstall I think it's finished - it does optimize `Thread.CurrentThread.ManagedThreadId` call sequence.

we don't need this optimization if we rewrite all `Thread.CurrentThread.ManagedThreadId` usages to `Environment.CurrentManagedThreadId` but looks like developers tend to prefer the former one.

**Comment by jkotas** [comment]
I have opened https://github.com/dotnet/runtime/issues/43257 on adding the analyzer. I think having analyzer for this is a good idea independ on whether we keep this as a pattern match optimization in the JIT.

**Comment by EgorBo** [comment]
So if we're fine with the analyzer (I filed a PR against aspnetcore to replace `Thread.CurrentThread.ManagedThreadId` with `Env.CurrentManagedThreadId`) I can just remove traces of the optimization in JIT or we can merge it as is (it works).

**Comment by AndyAyersMS** [comment]
I think we should probably handle this in the jit as there are likely many uses of  `Thread...` out there that might never get changed.

Can you resolve the conflicts?

File: src/coreclr/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs [hunk 1]
```diff
@@ -7,6 +7,7 @@
 using System.Runtime.InteropServices;
 using System.Runtime.Serialization;
 using System.Runtime.Versioning;
+using Internal.Runtime.CompilerServices;
```
Comment (jkotas) [inline src/coreclr/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs:10]: Why do we need this? `IntrinsicAttribute` is under `System.Runtime.CompilerServices`

Comment (EgorBo) [reply src/coreclr/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs]: Ah, right, removed.

**Review by AndyAyersMS — APPROVED** [review]
This looks ok on the jit side.

Perhaps @jkotas should also sign off?

File: src/coreclr/src/inc/corinfo.h [hunk 2]
```diff
@@ -208,11 +208,11 @@ TODO: Talk about initializing strutures before use
 //
 //////////////////////////////////////////////////////////////////////////////////////////////////////////
 
-constexpr GUID JITEEVersionIdentifier = { /* 062114d0-bd20-483f-8a3e-c4ee39706ae8 */
-    0x062114d0,
-    0xbd20,
-    0x483f,
-    {0x8a, 0x3e, 0xc4, 0xee, 0x39, 0x70, 0x6a, 0xe8}
+constexpr GUID JITEEVersionIdentifier = { /* a0184d06-a562-43f6-94ad-44627db87310 */
```
Comment (jkotas) [inline src/coreclr/src/inc/corinfo.h:211]: Would you mind deleting unused `CORINFO_HELP_INTERNALTHROW` and `CORINFO_HELP_SEC_UNMGDCODE_EXCPT` while you are bumping the JIT/EE interface GUID?

(I had it on my cleanup list, but I did want to bump the GUID just to delete a few helpers.)

**Review by jkotas — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**